### ISSUE TITLE
trigger timing for readyCallback and initCallback

### ISSUE THREAD

**Issue by wei2go** [davidjbradshaw/iframe-resizer#528]
Hi there,

Just wondering the timing for the initCallback on parent page and readyCallback on iframe page are triggered.

initCallback: doesn't seem to be triggered at all?

readyCallback: 
1. It's weird that occasionally it's not triggered. See the log attached.
2. It's triggered a bit slow (Does it happen on window.load? Can we have ability to define a setting and make it happen earlier, e.g. DOM ready?)

I don't have any special settings (all by default) for the iframe resizer.

Thanks,

Log (readyCallback not triggered for some reasons, can also tell the height calculation is incorrect):
```
[iFrameSizer][Host page: iFrameResizer0] Added missing iframe ID: iFrameResizer0 (...)
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] IFrame scrolling disabled for iFrameResizer0
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] [init] Sending msg to iframe[iFrameResizer0] (iFrameResizer0:8:false:true:32:true:true:null:bodyOffset:null:null:0:false:parent:scroll) targetOrigin: *
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Initialising iFrame (...)
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] TargetOrigin for parent set to: *
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] HTML & body height set to "auto"
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Enable public methods
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Add event listener: Animation Start
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Add event listener: Animation Iteration
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Add event listener: Animation End
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Add event listener: Input
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Add event listener: Mouse Up
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Add event listener: Mouse Down
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Add event listener: Orientation Change
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Add event listener: Print
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Add event listener: Ready State Change
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Add event listener: Touch Start
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Add event listener: Touch End
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Add event listener: Touch Cancel
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Add event listener: Transition Start
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Add event listener: Transition Iteration
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Add event listener: Transition End
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Create body MutationObserver
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] In page linking not enabled
my-page:1 WebSocket connection to 'wss://ws.pusherapp.com/app/edf8b9a8b28d1c28ba27?protocol=6&client=js&version=2.0.4&flash=false' failed: WebSocket is closed before the connection is established.
iframeResizer.min.js:8 [iFrameSizer][Host page: null] Ignored: {"messageId":0,"value":false}
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Trigger event lock on
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Sending message to host page (iFrameResizer0:15387:196:init)
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] Received: [iFrameSizer]iFrameResizer0:15387:196:init
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] Checking height is in range 0-Infinity
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] Checking width is in range 0-Infinity
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] IFrame (iFrameResizer0) height set to 15387px
custom.js:6 iFrame Resizer initialised
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Trigger event lock off
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] --
iframeResizer.min.js:8 [iFrameSizer][Host page: null] Ignored: {"cookies":{},"namespace":"twttr:cookies"}
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Trigger event: mutationObserver: [object HTMLScriptElement] childList
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] No change in size detected
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Trigger event: mutationObserver: [object HTMLBodyElement] childList
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] No change in size detected
5iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Trigger event: mutationObserver: [object HTMLBodyElement] childList
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] No change in size detected
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Trigger event: mutationObserver: [object HTMLBodyElement] childList
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] No change in size detected
3iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Trigger event: mutationObserver: [object HTMLBodyElement] childList
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Trigger event lock on
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Sending message to host page (iFrameResizer0:15397:200:mutationObserver)iFrameResizer0:15397:200:mutationObserver"defaultPrevented: falseeventPhase: 0isTrusted: truelastEventId: ""origin: "..."path: [Window]ports: []returnValue: truesource: Window {window: Window, self: Window, location: Location, closed: false, frames: Window, …}srcElement: Window {stop: ƒ, open: ƒ, alert: ƒ, confirm: ƒ, prompt: ƒ, …}target: Window {stop: ƒ, open: ƒ, alert: ƒ, confirm: ƒ, prompt: ƒ, …}timeStamp: 2287.9type: "message"__proto__: MessageEvent
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] Received: [iFrameSizer]iFrameResizer0:15397:200:mutationObserver
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] Checking height is in range 0-Infinity
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] Checking width is in range 0-Infinity
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] Requesting animation frame
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Trigger event: mutationObserver: [object HTMLDivElement] childList
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Sending message to host page (iFrameResizer0:15407:200:mutationObserver)
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] IFrame (iFrameResizer0) height set to 15397px
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] Received: [iFrameSizer]iFrameResizer0:15407:200:mutationObserver
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] Checking height is in range 0-Infinity
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] Checking width is in range 0-Infinity
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] Requesting animation frame
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] IFrame (iFrameResizer0) height set to 15407px
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Trigger event lock off
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] --
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Trigger event: mutationObserver: [object HTMLDivElement] attributes
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] No change in size detected
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] [iFrame.onload] Sending msg to iframe[iFrameResizer0] (iFrameResizer0:8:false:true:32:true:true:null:bodyOffset:null:null:0:false:parent:scroll) targetOrigin: *
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Trigger event: mutationObserver: [object HTMLBodyElement] childList
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] No change in size detected
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Trigger event: mutationObserver: [object HTMLBodyElement] childList
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Trigger event lock on
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Sending message to host page (iFrameResizer0:15406:200:mutationObserver)
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] Received: [iFrameSizer]iFrameResizer0:15406:200:mutationObserver
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] Checking height is in range 0-Infinity
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] Checking width is in range 0-Infinity
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] Requesting animation frame
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] IFrame (iFrameResizer0) height set to 15406px
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Trigger event lock off
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] --
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Trigger event: mutationObserver: [object HTMLDivElement] childList
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Trigger event lock on
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Sending message to host page (iFrameResizer0:16510:200:mutationObserver)
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] Received: [iFrameSizer]iFrameResizer0:16510:200:mutationObserver
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] Checking height is in range 0-Infinity
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] Checking width is in range 0-Infinity
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] Requesting animation frame
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] IFrame (iFrameResizer0) height set to 16510px
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Trigger event lock off
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] --
4iframeResizer.min.js:8 [iFrameSizer][Host page: null] Ignored: [object Object]
2iframeResizer.min.js:8 [iFrameSizer][Host page: null] Ignored: [object Object]
iframeResizer.min.js:8 [iFrameSizer][Host page: null] Ignored: [object Object]
iframeResizer.min.js:8 [iFrameSizer][Host page: null] Ignored: [object Object]
iframeResizer.min.js:8 [iFrameSizer][Host page: window] Trigger event: resize
iframeResizer.min.js:8 [iFrameSizer][Host page: iFrameResizer0] [Window resize] Sending msg to iframe[iFrameResizer0] (resize) targetOrigin: *
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] Trigger event: Parent window requested size check
iframeResizer.contentWindow.min.js:9 [iFrameSizer][iFrameResizer0] No change in size detected
```

**Comment by davidjbradshaw** [comment]
**initCallback** is called after the first message is received back from the iFrame.
**readyCallback** is call after the after the first message back to the parent page as been sent.

You seem to have other things posting to your parent page. Maybe they are causing an issue. Difficult to say from just the log.

timing is based on the `readystatechange` event firing.

**Comment by wei2go** [comment]
Hey David,

Thanks for your response.

I found the issue of readyCallback not getting fired sometimes is because I wrapped the handler in the dom.ready thing, and obviously sometimes the callback is fired earlier than my handler getting initialised. I then got it out of the wrap and put it above the iFrameResizer.contentWindow.js. All good now.

For the speeding issue, I haven't fully worked out why it takes so long for the iframe to load yet, but I  modified the iFrameResizer.contentWindow.js a bit, by removing the 'if('loading' !== document.readyState)' condition check to force it to send the ready message right after the script is loaded, and that seems to make the iframe load a lot faster. I'm not sure if this works for everybody, but I tested it and it doesn't seem to cause other issue so far.
